### PR TITLE
add task solution

### PR DESCRIPTION
**Pull request by opiquem**
[DEMO LINK](https://maksym-bobryk.github.io/react_add-todo-form/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,61 +1,135 @@
...
-          <span className="error">Please enter a title</span>
+          <label>
+            {'Title: '}
+
+            <input
+              type="text"
+              data-cy="titleInput"
+              value={title}
+              onChange={handleTitleChange}
+            />
```
Comment (33Orange) [inline src/App.tsx:88]: You don't have placeholder in title input

image
image

File: src/App.tsx [hunk 2]
```diff
@@ -1,61 +1,135 @@
...
+
+    setTitleValid(Boolean(title));
+    setUserValid(Boolean(selectedUser));
+
+    if (!selectedUser || !title) {
+      return;
+    }
+
+    setTodos((currentTodos) => {
+      const maxId = Math.max(...currentTodos.map(todo => todo.id));
```
Comment (33Orange) [inline src/App.tsx:51]: Create a separate function which will be creating new id and move it to separate file `helpers`

File: src/App.tsx [hunk 3]
```diff
@@ -1,61 +1,135 @@
...
+import usersFromServer from './api/users';
+import todosFromServer from './api/todos';
+
+const getUserById = (userId: number): User | null => (
+  usersFromServer.find(user => user.id === userId) || null
+);
+
+const getUserByName = (name: string): User | null => (
+  usersFromServer.find(user => user.name === name) || null
+);
```
Comment (33Orange) [inline src/App.tsx:16]: You can find user by id and you have this helper. 
Do you need this function?

File: src/App.tsx [hunk 4]
```diff
@@ -1,61 +1,135 @@
...
 import './App.scss';
 
-// import usersFromServer from './api/users';
-// import todosFromServer from './api/todos';
+import usersFromServer from './api/users';
+import todosFromServer from './api/todos';
+
+const getUserById = (userId: number): User | null => (
+  usersFromServer.find(user => user.id === userId) || null
+);
```
Comment (33Orange) [inline src/App.tsx:12]: Move it to separate file `helpers`

File: src/App.tsx [hunk 5]
```diff
@@ -1,61 +1,135 @@
...
+  const [todos, setTodos] = useState(refactoredTodos);
+  const [title, setTitle] = useState('');
+  const [selectedUser, setSelectedUser] = useState('');
+  const [titleValid, setTitleValid] = useState(true);
+  const [userValid, setUserValid] = useState(true);
+
+  const handleTitleChange = (event: ChangeEvent<HTMLInputElement>) => {
+    setTitle(event.target.value);
+    setTitleValid(Boolean(event.target.value));
+  };
```
Comment (33Orange) [inline src/App.tsx:33]: I can push todo with spaces in title

image

File: src/App.tsx [hunk 6]
```diff
@@ -1,61 +1,135 @@
...
+                value=""
+                disabled
+              >
+                Choose a user
+              </option>
+
+              {usersFromServer.map((user) => (
+                <option
+                  key={user.id}
+                  value={user.name}
```
Comment (33Orange) [inline src/App.tsx:114]: Use `userId` as value in select.
That helps you to remove function `getUserByName` and will be more correct.

File: src/App.tsx [hunk 7]
```diff
@@ -1,61 +1,135 @@
...
+    setTitleValid(Boolean(title));
+    setUserValid(Boolean(selectedUser));
+
+    if (!selectedUser || !title) {
+      return;
+    }
+
+    setTodos((currentTodos) => {
+      const maxId = Math.max(...currentTodos.map(todo => todo.id));
+      const newUser = getUserByName(selectedUser);
```
Comment (33Orange) [inline src/App.tsx:52]: Use ID as value in select and find user with your helper `getUserById`

**Review by 33Orange — CHANGES_REQUESTED** [review]
Almost done

A few moments to fix

File: src/App.tsx [hunk 8]
```diff
@@ -1,61 +1,128 @@
...
+    setTitleValid(Boolean(title.trim()));
+    setUserValid(Boolean(selectedUser));
+
+    if (!selectedUser || !title.trim()) {
+      return;
+    }
+
+    setTodos((currentTodos) => {
+      const nextId = getNextId(currentTodos);
+      const newUser = getUserById(selectedUser);
```
Comment (33Orange) [inline src/App.tsx:44]: Are you creating new user or just find exist ?

File: src/App.tsx [hunk 9]
```diff
@@ -1,61 +1,128 @@
...
+      return [
+        ...currentTodos,
+        {
+          id: nextId,
+          title,
+          completed: false,
+          userId: newUser ? newUser.id : null,
+          user: newUser,
+        },
+      ];
```
Comment (33Orange) [inline src/App.tsx:55]: Let's create new todo as new variable and just add it to array in setter to improve readability

```suggestion
   const newTodo = {
          id: getNextId(currentTodos),
          title,
          completed: false,
          userId: selectedUser? selectedUser : null,
          user: getUserById(selectedUser),
    };
 
    setTodos((currentTodos) => [...currentTodos, newTodo]);
```

**Review by 33Orange — CHANGES_REQUESTED** [review]
Almost done

File: src/App.tsx [hunk 10]
```diff
@@ -1,61 +1,124 @@
...
+  const [userValid, setUserValid] = useState(true);
+
+  const handleTitleChange = (event: ChangeEvent<HTMLInputElement>) => {
+    setTitle(event.target.value);
+    setTitleValid(Boolean(event.target.value));
+  };
+
+  const handleUserChange = (event: ChangeEvent<HTMLSelectElement>) => {
+    setSelectedUser(+event.target.value);
+    setUserValid(Boolean(event.target.value));
```
Comment (dffUqp) [inline src/App.tsx:29]:
```suggestion
    setUserValid(false);
```

**Review by dffUqp — APPROVED** [review]
LGTM:smiley:

**Pull request closed without merging**